[Merged by Bors] - feat(group_theory/noncomm_pi_coprod): homomorphism from pi monoids or groups

nomeata opened the pull request:
Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

---

@alreadydone wrote on zulip:

> It may be worth doing the following: given a family of homomorphisms Hi -> G into the same group, if the ranges of any two different homomorphisms commute, then we get a canonical homomorphism from a subgroup of the pi group of the Hi's (with finitely many components non-identity), or alternatively from a quotient group (by the commutators between different Hi's) of the free_product, into G. Then we can state that this canonical homomorphism is an isomorphism, and it will be a Prop.

This nerd-sniped me, and suddenly it is past 2am, but here it is:

* The construction of the homomorphism
* A lemma about its range
* Injectivity if the ranges are `complete_lattice.independent`
* The `mul_equiv` statement in the latter case.

I restricted myself to case of finitely many groups.

I plan to use this in [#nilpotent](https://github.com/leanprover-community/mathlib/issues/11723). But I expect this will need thorough cleanup:

* [ ] Where this should go
* [ ] namespacing and naming
* [ ] lots of code golfing,
* [x] maybe partial rewrite
* [x] documentation

So it’s a draft for now. But do fire away!

---

- [x] depends on: #12268
- [x] depends on: #12291
- [x] depends on: #12520
- [x] depends on: #12521
- [x] depends on: #12523
- [x] depends on: #12524
- [x] depends on: #12525

No longer a draft, but cleaned up and documented. Now less ashamed of it :-)

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

nomeata committed:
first stab at hom out of pi group

nomeata committed:
Please lint a bit

nomeata committed:
time flies

nomeata committed:
shuffling whitespace

nomeata committed:
Get rid of ugly have

eric-wieser commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,232 @@
+
+/-- TODO -/
+noncomputable
+def fun_on (S : finset I) : G := (list.map (λ i, ϕ i (f i)) S.to_list).prod
```
I think it's unprovable whether `fun_on ϕ f {i, j}` is `ϕ i (f i) * ϕ j (f j)` or `ϕ j (f j) * ϕ i (f i)`, right? If so, this API is a little weird.

nomeata replied:
It's the same, but I gotta start with something, don't I?

I consider this internal to the construction. Should I mark it as `private`?

It's inspired by https://leanprover-community.github.io/mathlib_docs/data/finset/noncomm_prod.html#finset.noncomm_prod but without a proof obligation that mentions `S`.

I could `include hcomm` for this definition (and not just the proofs for it) if it makes it less odd?

nomeata replied:
Hmm, maybe I can use `noncomm_prod` after all, if I hide the annoying proof argument in a definition like this `fun_on`, discharged via the `hcomm` assumption. Then I can simplify this stuff here a bit. Will try.

nomeata replied:
Interesting, `noncomm_prod` is `noncomputable`. That’s a good reason to use that…

nomeata replied:
I’ve ported it to use `noncomm_prod`. Benefits: Less to prove, and no `noncomputable`. 

Downside: The API for `noncomm_prof` insists using `insert a s` over `cons a s`. This requires `[decidable_eq I]`. The simp-setup of `finset` also rewrites `cons` to `insert`, but then lemmas like
```
lemma fun_on_insert_of_not_mem (S : finset I) (i : I) (h : i ∉ S) :
  fun_on ϕ f (insert i S) = ϕ i (f i) * fun_on ϕ f S :=
```
don’t work well as `simp` anymore, because the `h` is no longer mentioned on the LHS. Oh well.

nomeata committed:
Base development on noncomm_prod

nomeata committed:
Redundant [decidable_eq I]

nomeata committed:
Specialize the theorem to normal subgroups and subgroup inclusion

which is what I need for https://github.com/leanprover-community/mathlib/issues/11723

nomeata committed:
feat(group_theory/subgroup/basic): add commute_of_normal_of_disjoint

nomeata committed:
Stash coprimality argument

nomeata committed:
Merge remote-tracking branch 'origin/joachim/commute_of_normal' into joachim/pi_nom

eric-wieser commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,260 @@
+
+-- We assume that the elements of different morphism commute
+-- Since we need this all over the place we wrap it up in `fact`
+variables [hcomm : fact (∀ (i j : I), i ≠ j → ∀ (x : H i) (y : H j), commute (ϕ i x) (ϕ j y)) ]
```
Why do you need this? Is it just because you're using `noncomm_prod`? Is there a reason you can't just use `list.prod` and work with `list`s (possibly subtyped by `list.nodup`) instead?

nomeata replied:
No, I’ll need this assumption in any case, else I won’t be able to show that the function from the product into `G` is actually a homomorphism.

nomeata committed:
Injectivity via coprimality of groups. More suited for sylow group.

nomeata committed:
Big clean-up and documentation

nomeata committed:
Prune imports

nomeata committed:
Remove kinda unused lemma

nomeata committed:
whitespace

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,392 @@
+  intros i hi,
+  rw nat.is_coprime_iff_coprime,
+  apply h i hi,
+end
```
It seems that the useful lemma `is_coprime.prod_left` only exists for rings, not for `ℕ`, and the casting is surprisingly tricky. Maybe I am doing something wrong here?

Maybe we want all of the lemmas in https://leanprover-community.github.io/mathlib_docs/ring_theory/coprime/lemmas.html also about `nat.coprime`?

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,392 @@
+  apply nat.dvd_antisymm; rewrite order_of_dvd_iff_pow_eq_one,
+  { rw [inv_pow, pow_order_of_eq_one, one_inv], },
+  { nth_rewrite 0 ← (inv_inv x), rw [inv_pow, pow_order_of_eq_one, one_inv], }
+end
```
I’d be surprised if this doesn’t already exist, but I didn't find it.

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,392 @@
+lemma mul_eq_one_of_disjoint
+  {H₁ H₂ : subgroup G} (hdis : disjoint H₁ H₂) {x y : G} (hx : x ∈ H₁) (hy : y ∈ H₂)
+  (heq : x * y = 1) : x = 1 ∧ y = 1 :=
+begin
```
A possibly useful lemma for `disjoint` groups, but not that deep actually. Move there, keep here (with `private`), inline into its use below?

nomeata committed:
Ups

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,392 @@
+
+    have hij : i ≠ j, by {simp at hnmem, tauto},
+    have hiS : i ∉ S, by {simp at hnmem, tauto},
+    calc ϕ i (g i) * (ϕ j (f j) * (to_fun ϕ f S : G)) -- TODO: Why do I have to mention `ϕ`?
```
I thought by making `ϕ` a parameter I don’t have to mention it here again. So why do I?

nomeata replied:
Ok, known problem, tacic mode undoes parameters: https://github.com/leanprover/lean/issues/1773

nomeata replied:
```suggestion
    calc ϕ i (g i) * (ϕ j (f j) * (to_fun ϕ f S : G))
```

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,392 @@
+section family_of_groups
+
+-- We have an family of groups
+parameters {I : Type*} [fintype I] [decidable_eq I] {H : I → Type*} [∀ i, group (H i)]
```
Todo: check if `fintype` is only needed in `pi_hom` (the linter says so)

nomeata commented on src/group_theory/subgroup/basic.lean:
```diff
@@ -1228,6 +1228,14 @@ begin
+def _root_.group.comm_group_of_center_eq_top (h : center G = ⊤) : comm_group G :=
+{ mul_comm := by { rw eq_top_iff' at h, intros x y, exact h y x },
+  .. (_ : group G) }
+
```
TODO: did some git branch juggling mistake. Will fix.

nomeata committed:
Make it lint-clean

nomeata committed:
Less underscores

nomeata committed:
erge branch 'master' of github.com:leanprover-community/mathlib into joachim/pi_nom

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,401 @@
+  (heq : x * y = 1) : x = 1 ∧ y = 1 :=
+begin
+  have : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
+  subst this,
```
```suggestion
  obtain rfl : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
```

nomeata committed:
Update src/group_theory/pi_hom.lean

nomeata committed:
Proof cosmetics

nomeata committed:
Move order of inv

alreadydone commented on src/group_theory/order_of_element.lean:
```diff
@@ -317,6 +317,14 @@ begin
+  apply nat.dvd_antisymm; rewrite order_of_dvd_iff_pow_eq_one,
+  { rw [inv_pow, pow_order_of_eq_one, one_inv], },
+  { nth_rewrite 0 ← (inv_inv x), rw [inv_pow, pow_order_of_eq_one, one_inv], }
+end
```
```suggestion
by simp [order_of_eq_order_of_iff]
```

nomeata replied:
That’s a bit embarrassing for me :-)

alreadydone commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,384 @@
+
+Given a family of morphisms `ϕ i : H i →* G` for each `i : I` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
+`pi_hom.hom : (Π (i : I), H i) →* G` that coincides with `ϕ``
```
```suggestion
`pi_hom.hom : (Π (i : I), H i) →* G` that coincides with `ϕ`
```

alreadydone commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,384 @@
+
+## Main theorems
+
+* `pi_hom.hom` conicides with each `ϕ`
```
```suggestion
* `pi_hom.hom` conicides with `ϕ i` when restricted to `H i`
```

alreadydone commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,384 @@
+  intros i hi,
+  rw nat.is_coprime_iff_coprime,
+  apply h i hi,
+end
```
```suggestion
finset.prod_induction s (λ y, y.coprime x) (λ a b, nat.coprime.mul) (by simp)
```

alreadydone reviewed (commented):
These are for now and I'll continue the review tomorrow!

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,384 @@
+
+@[simp]
+lemma hom_update_one (i : I) (y : H i):
+  hom S (function.update 1 i y) = if i ∈ S then ϕ i y else 1 := to_fun_update_one _ _ _
```
Should use `mul_single` once https://github.com/leanprover-community/mathlib/pull/11849 lands

nomeata committed:
Apply suggestions from code review

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

nomeata committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into joachim/pi_nom

nomeata committed:
Use monoid_hom.single, not function.update

alreadydone commented on src/group_theory/order_of_element.lean:
```diff
@@ -317,6 +317,10 @@ begin
 
+@[simp, to_additive]
+lemma order_of_inv (x : G) : order_of x⁻¹ = order_of x :=
+by simp [order_of_eq_order_of_iff]
```
What do people think of making `order_of_eq_order_of_iff` an ext lemma?

kim-em replied:
No, that would be beyond the usual scope of `ext`, which is "please prove this equality by working 'componentwise'" for some generalised notion of 'componentwise'.

alreadydone committed:
simplify independence proof and remove lemmas

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -184,37 +184,8 @@ begin
-                  ... ∣ (∏ (i : I) in S, fintype.card (H i)) : finset.dvd_prod_of_mem _ hmem,
-end
+lemma pow (k : ℕ) : (hom S f) ^ k = hom S (f ^ k) := by simp
+/- uses map_pow, and now unnecessary -/
```
Silly me, for some reason I thought this doesn’t hold for all homomorphism. Thanks :-)

eric-wieser commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,325 @@
+
+## Main definitions
+
+* `pi_hom.hom : (Π (i : I), H i) →* G` is the main homomorphism
```
This is analogous to [`linear_map.lsum`](https://leanprover-community.github.io/mathlib_docs/linear_algebra/pi.html#linear_map.lsum), right?

nomeata replied:
It looks similar, yes

nomeata replied:
Although we don't have an equivalence here due to the extra requirement that elements from different ranges commute.

nomeata replied:
Maybe it would be better to put these definitions and lemmas into the `monoid_hom` namespace? But what would be a good name? 
Given that we have [monoid_hom.coprod](https://leanprover-community.github.io/mathlib_docs/algebra/group/prod.html#monoid_hom.coprod), probably `monoid_hom.copi` is the right name?

nomeata replied:
So how about

> * `monoid_hom.copi : (Π (i : I), N i) →* M` is the main homomorphism
> * `monoid_hom.copi_on (S : finset S) : (Π (i : I), N i) →* M` is the homomorphism restricted to the
>    set `S`, and mostly of internal interest in this file, to allow inductive proofs.
> * `subgroup.copi : (Π (i : I), H i) →* G` is the specialization to `H i : subgroup G` and
>    the subgroup embedding.

alreadydone replied:
I don't think any mathematician has the instinct to search for `copi` for such a thing. Maybe `comm_coprod` is a good name, as `pi_hom` is analogous to the "desc" field in [colimit](https://leanprover-community.github.io/mathlib_docs/category_theory/limits/is_limit.html#category_theory.limits.is_colimit) (coproduct in this case), except (the finite-supported) pi isn't exactly the coproduct (i.e. free product), but its quotient by the pairwise commutators.

nomeata replied:
But wouldn't `comm_coprod` be the one about `×` (if we'd add that), not the one about pi?

alreadydone replied:
In category theory "coproduct" isn't limited to binary, and is called [binary_coproduct](https://leanprover-community.github.io/mathlib_docs/category_theory/limits/shapes/types.html#category_theory.limits.types.binary_coproduct_cocone) when it's binary. Using `coprod` here may conflict with binary coproduct, but agrees with the usage in [finset.prod](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/basic.html#finset.prod). You could also start a [naming contest](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/naming.20contest) to gather opinions.

nomeata replied:
I don't care too strongly, my interest is mostly in reducing the size of the stack of unmerged PRs. So whatever someone with merge permissions (@eric-wieser ?) would be happy with is good for me :-)

nomeata replied:
The naming contest at https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Name.20for.20map.20from.20pi.20group was a bit inconclusive.

I’ll update the PR to go with `monoid_hom.noncomm_pi_coprod`, as suggested by @eric-wieser : The `noncomm_` prefix has precedence in [noncomm_prod](https://leanprover-community.github.io/mathlib_docs/data/finset/noncomm_prod.html#multiset.noncomm_prod), the `coprod` comes from [`monoid_hom.coprod`](https://leanprover-community.github.io/mathlib_docs/algebra/group/prod.html#monoid_hom.coprod) and the `pi` because `noncomm_coprod` would be for the binary product.

nomeata replied:
Is this better now?

nomeata committed:
Define hom for monoids

there is some unsatisfying redundancy wrt. `range` and `mrange`.

Also proved `mul_eq_one_iff_disjoint` to make this free of sorry.

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,364 @@
+  { rintros h x ⟨hx1, hx2⟩,
+    obtain rfl : x = 1 := (h hx1 (H₂.inv_mem hx2) (mul_inv_self x)).1,
+    exact rfl,
+  },
```
```suggestion
    exact rfl, },
```

nomeata committed:
Apply suggestions from code review

alreadydone committed:
polish proof

eric-wieser commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,360 @@
+
+-- We assume that the elements of different morphism commute
+-- Since we need this all over the place we wrap it up in `fact`
+parameters [hcomm : fact (∀ (i j : I), i ≠ j → ∀ (x : N i) (y : N j), commute (ϕ i x) (ϕ j y))]
```
Usually `fact` isn't for " I need this all over the place" but "I need this as an argument for something I plan to declare an `instance`".

nomeata replied:
Should I refactor and make this an explicit argument of the various definitions? (Maybe an implicit argument for the lemmas where it can be inferred)?

nomeata replied:
Remove the use of `fact`, and while I was it, also of `parameters`.

nomeata committed:
Don’t use fact nor parameters

nomeata committed:
Avoid simp followed by tauto

nomeata commented on src/group_theory/pi_hom.lean:
```diff
@@ -0,0 +1,370 @@
+/-!
+# Canonical homomorphism from a pi monoid
+
+This file defines txinput set-prop 11 "libinput Scroll Method Enabled" 0, 0, 1  #he construction of the canoncial homomorphism from a family of monoids.
```
```suggestion
This file defines the construction of the canoncial homomorphism from a family of monoids.
```

nomeata committed:
Update src/group_theory/pi_hom.lean

nomeata committed:
Say monoid_hom.noncomm_pi_coprod

eric-wieser commented on src/group_theory/order_of_element.lean:
```diff
@@ -136,6 +136,11 @@ is_periodic_pt.minimal_period_dvd ((is_periodic_pt_mul_iff_pow_eq_one _).mpr h)
 ⟨λ h, by rw [pow_eq_mod_order_of, nat.mod_eq_zero_of_dvd h, pow_zero], order_of_dvd_of_pow_eq_one⟩
 
+@[to_additive map_add_order]
+lemma map_order {H : Type*} [monoid H] (ψ : G →* H) (x : G) :
```
```suggestion
@[to_additive add_order_of_map_dvd]
lemma order_of_map_dvd {H : Type*} [monoid H] (ψ : G →* H) (x : G) :
```

eric-wieser reviewed (commented):
Can all of this be `to_additive`d?

nomeata committed:
Add to_additive

nomeata committed:
Update src/group_theory/order_of_element.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

nomeata committed:
Use order_of_map_dvd

nomeata committed:
Additive docstrings

nomeata committed:
s/I/ι/g

nomeata committed:
Fix typo in filename

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -1,6 +1,6 @@
 /-
 Copyright (c) 2022 Jocchim Breitner. All rights reserved.
-Released under Apache 2.0 license as described in the file LICENSE.
+Released under Apache 2.0 license as described in the file LιCENSE.
```
Err
```suggestion
Released under Apache 2.0 license as described in the file LICENSE.
```

nomeata replied:
🤦‍♂️

I should start using vim, I can't use vscode I guess

nomeata replied:
Yes, I failed to use the right replace command. Will fix.

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,402 @@
+* `subgroup.noncomm_pi_coprod_range`: The range of `subgroup.noncomm_pi_coprod` is `⨆ (i : ι), H i`.
+* `monoid_hom.injective_noncomm_pi_coprod_of_independent`: in the case of groups, `pi_hom.hom` is
+   injective if the `ϕ` are injective and the ranges of the `ϕ` are independent.
+* `monoid_hom.independent_range_of_coprime_order`: ιf the `N i` have coprime orders, then the ranges
```
```suggestion
* `monoid_hom.independent_range_of_coprime_order`: If the `N i` have coprime orders, then the ranges
```

You might want to check for any `ι`s followed by another letter...

nomeata committed:
Fix very embarrasing search-and-replace mistake

nomeata committed:
Refine some lemma names

nomeata committed:
Avoid open_locale classical

nomeata committed:
More pleasing classical linter

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+  {ι : Type*}
+  {x : ℕ} {s : ι → ℕ} {t : finset ι} :
+  (∀ (i : ι), i ∈ t → nat.coprime (s i) x) → nat.coprime (∏ (i : ι) in t, s i) x :=
+finset.prod_induction s (λ y, y.coprime x) (λ a b, nat.coprime.mul) (by simp)
```
Can you add this for `prod_right` too? (and find an appropriate file for both)

Up to you if you want to do it as a separate PR - although that will reduce your build time if this one undergoes a few more rounds of review.

nomeata replied:
https://github.com/leanprover-community/mathlib/pull/12268

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+end
+
+@[to_additive add_to_fun_in_sup_mrange]
+lemma to_fun_in_sup_mrange (S : finset ι) :
```
```suggestion
@[to_additive add_to_fun_mem_bsupr_mrange]
lemma to_fun_mem_bsupr_mrange (S : finset ι) :
```
`⨆ i ∈ S` is `bsupr`, `∈` is `mem`.

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+variable (hcomm)
+
+/-- The canonical homomorphism from a family of monoids. -/
+@[to_additive "The canonical homomorphism from a family of additive monoids."]
```
```suggestion
@[to_additive "The canonical homomorphism from a family of additive monoids.

See also `linear_map.lsum` for a linear version without the commutativity assumption."]
```

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+
+/-- The canonical homomorphism from a family of monoids. -/
+@[to_additive "The canonical homomorphism from a family of additive monoids."]
+def noncomm_pi_coprod : (Π (i : ι), N i) →* M := noncomm_pi_coprod_on.hom ϕ hcomm finset.univ
```
To match `linear_map.lsum`, is this true as an equiv?

```lean
def noncomm_pi_coprod_equiv :
  {ϕ : Π i, N i →* M // ∀ (i j : ι), i ≠ j → ∀ (x : N i) (y : N j), commute (ϕ i x) (ϕ j y) }
    ≃ ((Π i, N i) →* M) :=
```

nomeata replied:
Upon first glance, I think it is. I'll give it a shot!

nomeata replied:
I can show it if I have
```
@[to_additive]
lemma _root_.monoid_hom.pi_ext [decidable_eq ι] [fintype ι] {f g : (Π i, N i) →* M} :
  f = g ↔ (∀ i x, f (monoid_hom.single N i x) = g (monoid_hom.single N i x)) :=
sorry
```
which I yet have to prove, and find a home for, or find existing somewhere.

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+
+-- The subgroup version of `noncomm_pi_coprod_on.to_fun_in_sup_mrange`
+@[to_additive noncomm_pi_coprod_on.add_to_fun_in_sup_range]
+lemma noncomm_pi_coprod_on.to_fun_in_sup_range (S : finset ι) :
```
again, `bsupr` and `mem`

eric-wieser commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,426 @@
+
+namespace monoid_hom
+
+-- The subgroup version of `noncomm_pi_coprod_on_mrange`
```
`noncomm_pi_coprod_on_mrange` doesn't exist.

nomeata committed:
Move coprime_prod_left to src/data/nat/gcd.lean

nomeata committed:
feat(data/nat/gcd): add coprime_prod_left and coprime_prod_right

nomeata committed:
Merge branch 'joachim/coprie_prod_left' into joachim/pi_nom

nomeata committed:
s/in_sup/mem_bsupr/g

nomeata committed:
Review comments

nomeata committed:
Add universal property

nomeata committed:
More exploration

alreadydone commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,478 @@
+end
+
+@[to_additive]
+lemma _root_.monoid_hom.pi_ext [decidable_eq ι] [fintype ι] {f g : (Π i, N i) →* M}
```
I think it's best to use
https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/basic.html#monoid_hom.eq_of_eq_on_mdense
combined with `lemma mrange : (hom ϕ hcomm S).mrange = ⨆ i ∈ S, (ϕ i).mrange` applied to the inclusions of `H i` into the pi group (closre is defined in terms of Inf, and I believe ⨆ is also).
But if mrange is used then this can't be moved to the group.pi ... Finiteness of ι is also necessary here.

nomeata replied:
Hmm, I’m not sure if [`monoid_hom.eq_of_eq_on_mdense`](https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/basic.html#monoid_hom.eq_of_eq_on_mdense) saves me a lot – proof I have is already  quite short, building on a lemma stating that an element of `(x : Π i, N i)` is the product of it’s components if `ι` is finite. And that I have to prove anyways, whether in the form I have (as an explicit equality with `noncomm_prod`), or somehow phrased as saying that in  `(x : Π i, N i)` the closure of the `mrange`s of the `monoid_hom.single` is the whole group.

Maybe all of this should go into a new file `algebra.group.pi_finite`?

alreadydone replied:
Indeed, going through mdense seems convoluted, and your approach is more straightforward. `noncomm_prod_single` is the key lemma, and I'm now trying to fill in the `sorry`s by converting `noncomm_prod` to `list.prod`.

alreadydone replied:
Sorries killed! I think the first two lemmas could go into `noncomm_prod` while the other two could go into `group.pi` (importing `noncomm_prod`). I don't think it's worth creating a separate file for the finite case; when people have developed a multiplicative version of `finsupp` they may go back to generalize the current development to arbitrary index type.

nomeata replied:
Thanks! I generalized them a bit and will now move them into appropriate files and create a second PR.

nomeata replied:
> while the other two could go into group.pi (importing noncomm_prod)

It seems that `noncomm_prod` already imports `group.pi` indirectly, so I think they’ll go there as well

nomeata replied:
Ok, the yak has grown a few hairs again, to be shaved at https://github.com/leanprover-community/mathlib/pull/12291. I hope we don’t keep growing PRs at a fast rate than we can merge them :-)

alreadydone committed:
kill sorries

nomeata committed:
Classical linter

nomeata committed:
Generalize a bit

nomeata committed:
Simplify

nomeata committed:
Move lemmas

nomeata committed:
feat(data/list/prod_monoid): finite pi lemmas

including a few helpers

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
Update src/data/finset/noncomm_prod.lean

nomeata committed:
Rename to prod_le_pow_of_forall_le

nomeata committed:
Docstring

nomeata committed:
Don’ use monoid_hom.single_commute

nomeata committed:
Remove monoid_hom.single_commute

nomeata committed:
Generalize noncomm_prod_map and its dependencies

nomeata committed:
Line length

nomeata committed:
Fix fallout

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
More juggling of variables

nomeata committed:
Fix more fallout

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
More monoid_hom_class fallout

nomeata committed:
Undo changing `map_list_prod` to `monoid_hom_class`

nomeata committed:
Add docstrings

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into joachim/finite_pi_ext

nomeata committed:
Merge commit 'e96cf5e693dc14935c2cca01b091c09fd4bcc786' into joachim/coprie_prod_left

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
Merge branch 'joachim/coprie_prod_left' into joachim/pi_nom

nomeata committed:
Apply code review

nomeata committed:
Apply more code review

nomeata committed:
Phrase lemmas wrt. pi.mul_single

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
Fix apply-lemmas for monoid_hom.single

nomeata committed:
Fix apply-lemmas for monoid_hom.single

nomeata committed:
Stick closer to pi.mul_single

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
Fix fallout

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom
 und eine leere Beschreibung bricht den Commit ab.

nomeata committed:
Throw in some to_additive attributes

nomeata committed:
lint

nomeata committed:
lint

nomeata committed:
feat(data/list/prod_monoid): add prod_eq_pow_of_forall_eq

and rename `prod_le_of_forall_le` to `prod_le_pow_of_forall_le`

nomeata committed:
fix(src/algebra/group/pi): Fix apply-simp-lemmas for monoid_hom.single

so that the simp-normal form really is `pi.mul_single`.

While adjusting related lemmas in `group_theory.subgroup.basic`, add a
few missing `to_additive` attributes.

nomeata committed:
Merge branch 'joachim/single-simp-normal' into joachim/finite_pi_ext

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Use npow, not pow, in name, so that to_additive does the right thing

nomeata committed:
Rename le_prod_of_forall_le to npow_le_prod_of_forall_le

nomeata committed:
Use `∀ (x ∈ l),` syntax

nomeata committed:
Try to teach to_additive about `pow` naming convention

nomeata committed:
Explicit name to avoid name clashes

nomeata committed:
More names are correct now

nomeata committed:
Even more names are correct now

nomeata committed:
Even more names are correct now (order_of_element)

nomeata committed:
Even more names are correct now (exponent.lean)

nomeata committed:
Even more names are correct now (topology/continuous_function/algebra.lean)

nomeata committed:
Say pow, not npow

nomeata committed:
Merge commit '28c902d86d7904b662d6caf6c9b211b05c485728' into joachim/single-simp-normal

nomeata committed:
Merge branch 'joachim/single-simp-normal' into joachim/finite_pi_ext

nomeata committed:
Fix sum_divisors_prime_pow

nomeata committed:
More correct names

nomeata committed:
Merge branch 'joachim/to_additive_pow' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge commit 'eb46e7e3cf98edd45bd1cce2227edb90d7e4426d' into joachim/to_additive_pow

nomeata committed:
Merge branch 'joachim/to_additive_pow' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Merge branch 'joachim/finite_pi_ext' into joachim/pi_nom

nomeata committed:
Big refactoring, more lemmas about noncomm_prod, no more noncomm_pi_coprod_on

nomeata committed:
supr_le, not bsupr_le!

nomeata committed:
Weaken lemma

nomeata committed:
Make lint a bit happy

nomeata committed:
Reorder some hypotheses

nomeata committed:
Moving some lemmas

nomeata committed:
Rename lemma

nomeata committed:
feat(group_theory/subgroup/basic): disjoint_iff_mul_eq_one

nomeata committed:
feat(data/finset/noncomm_prod): assorted lemmas

* `noncomm_prod_congr`
* `noncomm_prod_commute`
* `noncomm_prod_mem_submonoid`
* `noncomm_prod_mem_subgroup`
* `noncomm_prod_mul_distrib`
* `eq_one_of_noncomm_prod_eq_one_of_independent`

nomeata committed:
Merge branch 'joachim/noncomm_prod_lemmas' into joachim/pi_nom

nomeata committed:
Update src/data/finset/noncomm_prod.lean

nomeata committed:
Merge branch 'joachim/noncomm_prod_lemmas' into joachim/pi_nom

nomeata committed:
Fix merge mistake

nomeata committed:
Avoid the need for finset.noncomm_prod_congr

nomeata committed:
Simplify further

nomeata committed:
Merge remote-tracking branch 'origin/master' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Undo renaming, now happening in #12589

nomeata committed:
Rename the new lemma

nomeata committed:
Merge commit 'e3ad468bb46b2773596ec7329c4a298d71a1df08' into joachim/prod_eq_pow_of_forall_eq

nomeata committed:
Merge branch 'joachim/prod_eq_pow_of_forall_eq' into joachim/finite_pi_ext

nomeata committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into joachim/finite_pi_ext

nomeata committed:
Rename lemmas according to latest fashion

nomeata committed:
Multisetify noncomm_prod_map

nomeata committed:
Multisetify noncomm_prod_eq_pow_card

nomeata committed:
Golf

nomeata committed:
Avoid rather specialized lemma

nomeata committed:
Use pi.mul_single_apply_commute

nomeata committed:
Apply suggestions from code review

Co-authored-by: Johan Commelin <johan@commelin.net>

nomeata committed:
Merge commit 'refs/pull/12291/head' of github.com:leanprover-community/mathlib into joachim/finite_pi_ext

nomeata committed:
Merge commit '09ea7fb00a1154b17423e8285313c61e7c427ad7' into joachim/finite_pi_ext

nomeata committed:
Post-merge update

kim-em commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,275 @@
+/-
+Copyright (c) 2022 Jocchim Breitner. All rights reserved.
```
```suggestion
Copyright (c) 2022 Joachim Breitner. All rights reserved.
```

nomeata replied:
:-D

kim-em commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,275 @@
+/-!
+# Canonical homomorphism from a finite family of monoids
+
+This file defines the construction of the canoncial homomorphism from a family of monoids.
```
```suggestion
This file defines the construction of the canonical homomorphism from a family of monoids.
```

kim-em commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,275 @@
+This file defines the construction of the canoncial homomorphism from a family of monoids.
+
+Given a family of morphisms `ϕ i : N i →* M` for each `i : ι` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
```
```suggestion
images of different morphisms commute, we obtain a canonical morphism
```

kim-em commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,275 @@
+
+## Main theorems
+
+* `monoid_hom.noncomm_pi_coprod` conicides with `ϕ i` when restricted to `N i`
```
```suggestion
* `monoid_hom.noncomm_pi_coprod` coincides with `ϕ i` when restricted to `N i`
```

kim-em commented on src/group_theory/noncomm_pi_coprod.lean:
```diff
@@ -0,0 +1,275 @@
+   injective if the `ϕ` are injective and the ranges of the `ϕ` are independent.
+* `monoid_hom.independent_range_of_coprime_order`: If the `N i` have coprime orders, then the ranges
+   of the `ϕ` are independent.
+* `subgroup.independent_of_coprime_order`: If commuting, normal subgroups `H i` have coprime orders,
```
```suggestion
* `subgroup.independent_of_coprime_order`: If commuting normal subgroups `H i` have coprime orders,
```

nomeata committed:
fix partly embarrassing typos

Co-authored-by: Scott Morrison <scott@tqft.net>
